duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -93,6 +93,45 @@ duk_ret_t javaObjectFinalizer(duk_context *ctx) {
   return 0;
 }
 
+void handleJavaScriptError(JNIEnv* env, duk_context* ctx) {
+  jclass exceptionClass = env->FindClass("com/squareup/duktape/DuktapeException");
+
+  // If it's a Duktape error object, try to pull out the full stacktrace.
+  if (duk_is_error(ctx, -1) && duk_get_prop_string(ctx, -1, "stack")) {
+    const char* stack = duk_safe_to_string(ctx, -1);
+
+    // Is there an exception thrown from a Java method?
+    if (duk_get_prop_string(ctx, -2, JavaMethod::JAVA_EXCEPTION_PROP_NAME)) {
+      jthrowable ex = static_cast<jthrowable>(duk_get_pointer(ctx, -1));
+
+      // add the Duktape JavaScript stack to this exception.
+      const jmethodID addDuktapeStack =
+          env->GetStaticMethodID(exceptionClass,
+                                 "addDuktapeStack",
+                                 "(Ljava/lang/Throwable;Ljava/lang/String;)V");
+      env->CallStaticVoidMethod(exceptionClass, addDuktapeStack, ex, env->NewStringUTF(stack));
+
+      // Rethrow the Java exception.
+      env->Throw(ex);
+
+      // Pop the Java throwable.
+      duk_pop(ctx);
+    } else {
+      env->ThrowNew(exceptionClass, stack);
+    }
+    // Pop the stack text.
+    duk_pop(ctx);
+  } else {
+    // Not an error or no stacktrace, just convert to a string.
+    env->ThrowNew(exceptionClass, duk_safe_to_string(ctx, -1));
+  }
+}
+
+void throwNoSuchMethod(JNIEnv* env, const std::string& message) {
```

Comment:
Nice. Java defines two exceptions that are almost identical here:

NoSuchMethodException: reflection failures
NoSuchMethodError: mismatch between compiler and code

I think you picked the right one. NoSuchMethodException is more common, but it’s a checked exception and you definitely need an unchecked exception.